streams/src/main/java/org/apache/kafka/streams/processor/internals/ProcessorStateManager.java

```diff
@@ -325,6 +334,12 @@ public void checkpoint(final Map<TopicPartition, Long> ackedOffsets) {
         }
         // write the checkpoint file before closing, to indicate clean shutdown
         try {
+            if (checkpoint == null) {
+                assert eosEnabled;
+                checkpoint = new OffsetCheckpoint(new File(baseDir, CHECKPOINT_FILE_NAME));
+            } else {
+                assert !eosEnabled;
```

Comment:
Assertions are enabled when we run tests  -- thus, we can catch bugs using them. I know we don't use them often, but I think we should use them everywhere: it helps to keep the control flow at a narrow pass, simplifying reasoning about it. I also like them a lot, as they also tell the reader of the code about invariants/assumptions that must hold.

Of course, we should only use assumption for invariants to check our own code -- we should not replace checks that must be user facing exceptions by asserts.

And for real deployment, they are disabled and thus there is no runtime overhead (just some startup overhead as the code get's removed AFAIK).